pychromecast/__init__.py

```diff
@@ -334,6 +344,33 @@ def join(self, timeout=None):
         """
         self.socket_client.join(timeout=timeout)
 
+    def start_dynamic_host_tracking(self):
+        """
+        Start dynamic non-blocking host tracking
+        """
+        self.tracker, self.browser = \
+            start_discovery(self._dynamic_host_callback)
+
+    def _dynamic_host_callback(self, unit_id):
+        """
+        Callback for dynamic discovery of host ip changes
+        """
+        unit = self.tracker.services[unit_id]
+        host = unit[0]
+        port = unit[1]
+        name = unit[4]
+
+        if not name == self.name or not port == self.port:
+            return  # Wrong group, keep discovering
+
+        if host == self.host:
+            return  # No change in host, keep discovering
+
+        self.logger.info('Change in host ip detected, '
+                         'reinitializing object...')
+        stop_discovery(self.browser)
+        self.__init__(host, port=port, device=self.device, dynamic_host=True)
```

Comment:
We definitely don't want to re-run `__init__` when the host changes. Beyond being bad practice this is going to have a lot of unintended side-effects, such as dropping all the old kwargs passed in.

What we'll want to do is pull everything from init that needs to be re-run when the host changes into a separate method, and call that method in init, and here. ([Probably everything from here down.](https://github.com/balloob/pychromecast/blob/master/pychromecast/__init__.py#L174))

You'll also need to clean-up the old socket client before creating the new one.